objects/src/transaction/proven_tx.rs

```diff
@@ -0,0 +1,111 @@
+            consumed_note_info,
+            created_note_info,
+            tx_script_root,
+            block_ref,
+            proof,
+        }
+    }
+
+    /// Verify the transaction using the provided data and proof.
+    pub fn verify(&self) -> Result<u32, VerificationError> {
+        // create verification dependencies
+        let stack_inputs = self.stack_inputs();
+        let stack_outputs = self.stack_outputs();
+        let program_info = ProgramInfo::new(self.tx_script_root(), Default::default());
+
+        // verify the transaction
+        verify(program_info, stack_inputs, stack_outputs, self.proof.clone())
+    }
+
+    // HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    fn stack_inputs(&self) -> StackInputs {
+        let mut stack_inputs: Vec<Felt> = Vec::with_capacity(13);
+        stack_inputs.extend_from_slice(self.block_ref.as_elements());
+        stack_inputs.push(*self.account_id);
+        stack_inputs.extend_from_slice(self.initial_account_hash.as_elements());
+        stack_inputs.extend_from_slice(self.consumed_notes_hash().as_elements());
+        stack_inputs.reverse(); //TODO: confirm this reversal is correct
+        StackInputs::new(stack_inputs)
+    }
+
+    fn stack_outputs(&self) -> StackOutputs {
+        // create the stack outputs
+        let mut stack_outputs: Vec<Felt> = Vec::with_capacity(8);
+        stack_outputs.extend_from_slice(self.final_account_hash.as_elements());
+        stack_outputs.extend_from_slice(self.created_notes_hash().as_elements());
+        stack_outputs.reverse(); //TODO: confirm this reversal is correct
+        StackOutputs::from_elements(stack_outputs, Default::default())
+    }
+
+    /// Returns the hash of the consumed notes.
+    fn consumed_notes_hash(&self) -> Digest {
+        // collect consumed note info into a single vector of elements and hash it
+        let mut elements: Vec<Felt> = Vec::with_capacity(self.consumed_note_info.len() * 8);
+        for note in self.consumed_note_info.iter() {
+            elements.extend_from_slice(note.nullifier().as_elements());
+            elements.extend_from_slice(note.script_root().as_elements());
+        }
+        Hasher::hash_elements(&elements)
+    }
+
+    /// Returns the hash of the created notes.
+    fn created_notes_hash(&self) -> Digest {
+        // collect created note info into a single vector of elements and hash it
+        let mut elements: Vec<Felt> = Vec::with_capacity(self.created_note_info.len() * 8);
+        for note in self.created_note_info.iter() {
+            elements.extend_from_slice(note.note_hash().as_elements());
+            elements.extend_from_slice(&note.note_metadata());
+        }
+        Hasher::hash_elements(&elements)
+    }
+
+    /// Returns the script root of the transaction.
+    fn tx_script_root(&self) -> Digest {
+        let _script_root = self.tx_script_root;
+        todo!()
```

Comment:
I think this can just return `self.tx_script_root` - no need for `todo!()` here.

Reply:
updated